From 684486b0896b97867eb5ccc181893147f38b3f44 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Fri, 12 Nov 2021 10:58:43 +0900
Subject: [PATCH 144/196] Add `IsExodusEnabled` (#231)

Checks whether the Exodus upgrade (at 2021-12-21T15:59:00Z) has activated yet, based on MTP.

Also replaces axionactivationtime with exodusactivationtime.
---
 src/consensus/activation.cpp                  | 14 ++++++-
 src/consensus/activation.h                    |  9 ++++-
 src/init.cpp                                  |  4 +-
 src/test/CMakeLists.txt                       |  2 +-
 src/test/activation_tests.cpp                 | 39 ++++++++++++++++++-
 test/CMakeLists.txt                           |  4 +-
 .../test_framework/test_framework.py          | 12 +++---
 test/lint/check-doc.py                        |  2 +-
 8 files changed, 70 insertions(+), 16 deletions(-)

diff --git a/src/consensus/activation.cpp b/src/consensus/activation.cpp
index 536a1b290..425509aaa 100644
--- a/src/consensus/activation.cpp
+++ b/src/consensus/activation.cpp
@@ -4,4 +4,16 @@
 
 #include <consensus/activation.h>
 
-/* Place future activation logic here. */
+#include <chain.h>
+#include <consensus/params.h>
+#include <util/system.h>
+
+bool IsExodusEnabled(const Consensus::Params &params,
+                     const CBlockIndex *pindexPrev) {
+    if (pindexPrev == nullptr) {
+        return false;
+    }
+
+    return pindexPrev->GetMedianTimePast() >=
+           gArgs.GetArg("-exodusactivationtime", params.exodusActivationTime);
+}
diff --git a/src/consensus/activation.h b/src/consensus/activation.h
index dcf6d5443..2e518b491 100644
--- a/src/consensus/activation.h
+++ b/src/consensus/activation.h
@@ -5,7 +5,14 @@
 #ifndef BITCOIN_CONSENSUS_ACTIVATION_H
 #define BITCOIN_CONSENSUS_ACTIVATION_H
 
+class CBlockIndex;
 
-/* Place future activation logic here. */
+namespace Consensus {
+struct Params;
+}
+
+/** Check if December 21st, 2021 protocol upgrade has activated. */
+bool IsExodusEnabled(const Consensus::Params &params,
+                     const CBlockIndex *pindexPrev);
 
 #endif // BITCOIN_CONSENSUS_ACTIVATION_H
diff --git a/src/init.cpp b/src/init.cpp
index 20e47ac06..f1f4080b9 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -418,8 +418,8 @@ void SetupServerArgs(NodeContext &node) {
         "-allowselfsignedrootcertificates", "-choosedatadir", "-lang=<lang>",
         "-min", "-resetguisettings", "-rootcertificates=<file>", "-splash",
         "-uiplatform",
-        // TODO remove after the November 2020 upgrade
-        "-axionactivationtime"};
+        // TODO remove after the December 2021 upgrade
+        "-exodusactivationtime"};
 
     // Set all of the args and their help
     // When adding new options to the categories, please keep and ensure
diff --git a/src/test/CMakeLists.txt b/src/test/CMakeLists.txt
index ad1577152..ed7791eea 100644
--- a/src/test/CMakeLists.txt
+++ b/src/test/CMakeLists.txt
@@ -276,7 +276,7 @@ function(add_boost_test_runners_with_upgrade_activated SUITE EXECUTABLE)
 			# Dec. 1st, 2019 at 00:00:00
 			--
 			"-testsuitename=Bitcoin ABC unit tests with next upgrade activated"
-			-axionactivationtime=1575158400
+			-exodusactivationtime=1575158400
 		)
 	endforeach()
 endfunction()
diff --git a/src/test/activation_tests.cpp b/src/test/activation_tests.cpp
index 0068addea..f54918d79 100644
--- a/src/test/activation_tests.cpp
+++ b/src/test/activation_tests.cpp
@@ -2,14 +2,49 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <chain.h>
+#include <chainparams.h>
+#include <consensus/activation.h>
+#include <util/system.h>
+
 #include <test/util/setup_common.h>
 
 #include <boost/test/unit_test.hpp>
 
 BOOST_FIXTURE_TEST_SUITE(activation_tests, BasicTestingSetup)
 
-BOOST_AUTO_TEST_CASE(activationtest) {
-    /* Place future activation logic tests here. */
+static void SetMTP(std::array<CBlockIndex, 12> &blocks, int64_t mtp) {
+    size_t len = blocks.size();
+
+    for (size_t i = 0; i < len; ++i) {
+        blocks[i].nTime = mtp + (i - (len / 2));
+    }
+
+    BOOST_CHECK_EQUAL(blocks.back().GetMedianTimePast(), mtp);
+}
+
+BOOST_AUTO_TEST_CASE(isexodusenabled) {
+    const Consensus::Params &params = Params().GetConsensus();
+    const auto activation =
+        gArgs.GetArg("-exodusactivationtime", params.exodusActivationTime);
+    SetMockTime(activation - 1000000);
+
+    BOOST_CHECK(!IsExodusEnabled(params, nullptr));
+
+    std::array<CBlockIndex, 12> blocks;
+    for (size_t i = 1; i < blocks.size(); ++i) {
+        blocks[i].pprev = &blocks[i - 1];
+    }
+    BOOST_CHECK(!IsExodusEnabled(params, &blocks.back()));
+
+    SetMTP(blocks, activation - 1);
+    BOOST_CHECK(!IsExodusEnabled(params, &blocks.back()));
+
+    SetMTP(blocks, activation);
+    BOOST_CHECK(IsExodusEnabled(params, &blocks.back()));
+
+    SetMTP(blocks, activation + 1);
+    BOOST_CHECK(IsExodusEnabled(params, &blocks.back()));
 }
 
 BOOST_AUTO_TEST_SUITE_END()
diff --git a/test/CMakeLists.txt b/test/CMakeLists.txt
index f9eb1d734..01ad338b0 100644
--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -96,14 +96,14 @@ add_dependencies(check-extended check-functional-extended)
 
 add_functional_test_check(check-functional-upgrade-activated
 	"functional tests with the next upgrade activated"
-	--with-axionactivation
+	--with-exodusactivation
 )
 add_dependencies(check-upgrade-activated check-functional-upgrade-activated)
 
 add_functional_test_check(check-functional-upgrade-activated-extended
 	"extended functional tests with the next upgrade activated"
 	--extended
-	--with-axionactivation
+	--with-exodusactivation
 )
 add_dependencies(check-upgrade-activated-extended check-functional-upgrade-activated-extended)
 
diff --git a/test/functional/test_framework/test_framework.py b/test/functional/test_framework/test_framework.py
index 97ca7778f..32f44daaa 100755
--- a/test/functional/test_framework/test_framework.py
+++ b/test/functional/test_framework/test_framework.py
@@ -178,8 +178,8 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
                             help="run nodes under the valgrind memory error detector: expect at least a ~10x slowdown, valgrind 3.14 or later required")
         parser.add_argument("--randomseed", type=int,
                             help="set a random seed for deterministically reproducing a previous test run")
-        parser.add_argument("--with-axionactivation", dest="axionactivation", default=False, action="store_true",
-                            help="Activate axion update on timestamp {}".format(TIMESTAMP_IN_THE_PAST))
+        parser.add_argument("--with-exodusactivation", dest="exodusactivation", default=False, action="store_true",
+                            help="Activate exodus update on timestamp {}".format(TIMESTAMP_IN_THE_PAST))
         parser.add_argument(
             '--timeout-factor',
             dest="timeout_factor",
@@ -455,9 +455,9 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
                 start_perf=self.options.perf,
                 use_valgrind=self.options.valgrind,
             ))
-            if self.options.axionactivation:
+            if self.options.exodusactivation:
                 self.nodes[i].extend_default_args(
-                    ["-axionactivationtime={}".format(TIMESTAMP_IN_THE_PAST)])
+                    ["-exodusactivationtime={}".format(TIMESTAMP_IN_THE_PAST)])
 
     def start_node(self, i, *args, **kwargs):
         """Start a lotusd"""
@@ -688,9 +688,9 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
                     emulator=self.options.emulator,
                 ))
 
-            if self.options.axionactivation:
+            if self.options.exodusactivation:
                 self.nodes[CACHE_NODE_ID].extend_default_args(
-                    ["-axionactivationtime={}".format(TIMESTAMP_IN_THE_PAST)])
+                    ["-exodusactivationtime={}".format(TIMESTAMP_IN_THE_PAST)])
 
             self.start_node(CACHE_NODE_ID)
             cache_node = self.nodes[CACHE_NODE_ID]
diff --git a/test/lint/check-doc.py b/test/lint/check-doc.py
index ef7b50fbc..86a24668a 100755
--- a/test/lint/check-doc.py
+++ b/test/lint/check-doc.py
@@ -51,7 +51,7 @@ SET_FALSE_POSITIVE_UNDOCUMENTED = set([
     # Removed arguments that now just print a helpful error message
     '-zapwallettxes',
     # Remove after November 2020 upgrade
-    '-axionactivationtime',
+    '-exodusactivationtime',
     '-replayprotectionactivationtime',
 ])
 
-- 
2.39.5 (Apple Git-154)

